x/examples/test-connectivity/main.go

```diff
@@ -200,43 +254,7 @@ func main() {
 					if err != nil {
 						return nil, err
 					}
-					var dnsStart time.Time
-					ctx = httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
-						DNSStart: func(di httptrace.DNSStartInfo) {
-							dnsStart = time.Now()
-						},
-						DNSDone: func(di httptrace.DNSDoneInfo) {
-							report := dnsReport{
-								QueryName:  hostname,
-								Time:       dnsStart.UTC().Truncate(time.Second),
-								DurationMs: time.Since(dnsStart).Milliseconds(),
-							}
-							if di.Err != nil {
-								report.Error = di.Err.Error()
-							}
-							for _, ip := range di.Addrs {
-								report.AnswerIPs = append(report.AnswerIPs, ip.IP.String())
-							}
-							// TODO(fortuna): Use a Mutex.
-							dnsReports = append(dnsReports, report)
-						},
-						ConnectDone: func(network, addr string, connErr error) {
-							ip, port, err := net.SplitHostPort(addr)
-							if err != nil {
-								return
-							}
-							report := tcpReport{
-								Hostname: hostname,
-								IP:       ip,
-								Port:     port,
-							}
-							if connErr != nil {
-								report.Error = connErr.Error()
-							}
-							// TODO(fortuna): Use a Mutex.
-							tcpReports = append(tcpReports, report)
-						},
-					})
+					ctx = getReportFromTrace(ctx, r, hostname)
```

Comment:
Both of our codes have a race condition. We need a mutex to append to the DNS queries and TCP connections.

I was imagining a different function here, one that takes callbacks with the dns and tcp reports instead, and the function can then append to whatever we need. That way the function doesn't depend on the specific report format.

BTW, it shouldn't depend on r, because the scope is too broad. It only needs the dns and tcp reports.

Reply:
@fortuna yes the race condition issue stands. I was planning to do that next. I can reduce the dependency on the report format